### PR TITLE
Extend PickTableLayout to pick layouts before add exchanges

### PR DESCRIPTION
**Pull request by rschlussel**
This adds what was in #8658 to the existing PickTableLayout rule.  This work needed for making cost based decisions.  We need to take our best guess at choosing a table layout before AddExchanges so that we can have more accurate statistics.  

supersedes #8658

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java [hunk 1]
```diff
@@ -235,6 +235,9 @@ public PlanOptimizers(Metadata metadata, SqlParser sqlParser, FeaturesConfig fea
                 new TransformCorrelatedSingleRowSubqueryToProject(),
                 new CheckSubqueryNodesAreRewritten(),
                 new PredicatePushDown(metadata, sqlParser),
+                new IterativeOptimizer(
+                        stats,
+                        new PickTableLayout(metadata, sqlParser, true).rules()),
```
Comment (rschlussel) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java]: I didn't bother updating the deprecated PickLayout rule too. We only need this change when other IterativeOptimizers are also on anyway.

**Comment by rschlussel** [comment]
seems to be causing problems with partitioned tables. I'm investigating.

**Comment by rschlussel** [comment]
I was able to reproduce the partitioned tables issue on master with a slightly different query.  This patch still can't be merged until that's resolved since we'll now get a test failure, but it's not directly related.  This is ready for review.

**Comment by rschlussel** [comment]
This is the partitioned tables issue: https://github.com/prestodb/presto/issues/8743

**Comment by kokosing** [comment]
Please ping me when the underlying issue is solved (merged) and this one get ready to be reviewed.

**Comment by rschlussel** [comment]
@kokosing This is ready for review again.  Thanks!

File: presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java (hunk 1 above)
Comment (kokosing) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java]: Can you please extract variable and use it in both places.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PickTableLayout.java [hunk 2]
```diff
@@ -55,10 +55,14 @@ public PickTableLayout(Metadata metadata)
...
         private final Metadata metadata;
+        private final SqlParser sqlParser;
+        private final boolean replaceExistingLayout;
 
-        private PickTableLayoutForPredicate(Metadata metadata)
+        private PickTableLayoutForPredicate(Metadata metadata, SqlParser sqlParser, boolean replaceExistingLayout)
         {
             this.metadata = requireNonNull(metadata, "metadata is null");
+            this.replaceExistingLayout = replaceExistingLayout;
+            this.sqlParser = sqlParser;
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PickTableLayout.java]: use require non null, here and always in constructor (unless it is some exceptional case)

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TableLayoutRewriter.java [hunk 3]
```diff
@@ -50,16 +52,18 @@
...
+    private final SqlParser sqlParser;
 
-    public TableLayoutRewriter(Metadata metadata, Session session, SymbolAllocator symbolAllocator, PlanNodeIdAllocator idAllocator)
+    public TableLayoutRewriter(Metadata metadata, Session session, SymbolAllocator symbolAllocator, PlanNodeIdAllocator idAllocator, SqlParser sqlParser)
     {
         this.metadata = metadata;
         this.session = session;
         this.symbolAllocator = symbolAllocator;
         this.idAllocator = idAllocator;
+        this.sqlParser = sqlParser;
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TableLayoutRewriter.java]: requireNonNull

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PickTableLayout.java [hunk 4]
```diff
@@ -38,11 +38,11 @@
 {
     private final ImmutableSet<Rule<?>> rules;
 
-    public PickTableLayout(Metadata metadata)
+    public PickTableLayout(Metadata metadata, SqlParser sqlParser, boolean replaceExistingLayout)
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PickTableLayout.java]: Can we remove this boolean parameter? I would prefer to have it set to false as it used to be so far, but this would require this rule to be run once. Alternatively, we can evaluate the rule always and rely on `planChanged` method.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java (hunk 1 above)
Comment (kokosing) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java]: Can you please explain me why do you execute it here and then why do you overwrite couple of optimizers later? Maybe we could run it once.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java [hunk 5]
```diff
@@ -269,6 +272,9 @@ public PlanOptimizers(Metadata metadata, SqlParser sqlParser, FeaturesConfig fea
                         ImmutableList.of(new com.facebook.presto.sql.planner.optimizations.EliminateCrossJoins()), // This can pull up Filter and Project nodes from between Joins, so we need to push them down again
                         ImmutableSet.of(new EliminateCrossJoins())),
                 new PredicatePushDown(metadata, sqlParser),
+                new IterativeOptimizer(
+                        stats,
+                        new PickTableLayout(metadata, sqlParser, true).rules()),
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java]: Is it needed here? Shortly `AddExchanges` will set a proper table layout, in single node mode you do not overwrite it anyway.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TableLayoutRewriter.java [hunk 6]
```diff
@@ -50,16 +52,18 @@
...
     }
 
-    public PlanNode planTableScan(TableScanNode node, Expression predicate)
+    public Optional<PlanNode> planTableScan(TableScanNode node, Expression predicate)
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TableLayoutRewriter.java]: I don't think it should be `TableLayoutRewriter` responsibility to check that the plan has changed. It should be the rule that checks it since it is required in order to not loop forever. Other callers of `planTableScan` don't care if plan changed.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TableLayoutRewriter.java [hunk 7]
```diff
@@ -100,11 +104,15 @@ public PlanNode planTableScan(TableScanNode node, Expression predicate)
...
         if (!BooleanLiteral.TRUE_LITERAL.equals(resultingPredicate)) {
-            return new FilterNode(idAllocator.getNextId(), result, resultingPredicate);
+            result = new FilterNode(idAllocator.getNextId(), rewrittenTableScan, resultingPredicate);
         }
 
-        return result;
+        if (result instanceof FilterNode && !planChanged(node, rewrittenTableScan, predicate, resultingPredicate)) {
+            return Optional.empty();
+        }
+        return Optional.of(result);
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TableLayoutRewriter.java]: You don't check if plan has changed if there are no filters, but just table scan

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TableLayoutRewriter.java [hunk 8]
```diff
@@ -115,4 +123,14 @@ public PlanNode planTableScan(TableScanNode node, Expression predicate)
                     || layout.getLayout().getColumns().get().containsAll(columnHandles);
         };
     }
+
+    private boolean planChanged(TableScanNode oldTableScan, TableScanNode rewrittenTableScan, Expression oldPredicate, Expression rewrittenPredicate)
+    {
+        if (!new ExpressionEquivalence(metadata, sqlParser).areExpressionsEquivalent(session, oldPredicate, rewrittenPredicate, symbolAllocator.getTypes())) {
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TableLayoutRewriter.java]: Do we need full `ExpressionEquivalence` here? Symbol names should be the same. Unless `resultingPredicate` is different each time `pickLayout` is called, then I think we can compare expressions directly.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java (hunk 1 above)
Comment (rschlussel) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java]: I run it after every predicate push down because if there's a new filter above the tablescan we may be able to choose a different layout.  And we want to make sure we're choosing the best layout we can so we can get good stats for any cost-based rules that run.  Even if right now there isn't anything that uses it in between the two times, cost-based optimizers can go anywhere, so we want the layout to be correct no matter where you put them.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java (hunk 5 above)
Comment (rschlussel) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java]: join enumeration would go after this.  We need the right table layout.

[the remark this answers is not in the document]
Comment (rschlussel) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java]: 1. the rule runs once earlier 
2. I don't think we should pay attention to whether the rule was or was not run earlier.  Because someone could always put another instance of the rule earlier, and then setting this to false would be incorrect.  The point is that in this invocation of the rule we don't care whether a table layout has been set.  We always want to replace the layout.  This is different from the invocation after AddExchanges.  AddExchanges always chooses the "right" layout, so if there's a layout set, we don't want to overwrite it.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PickTableLayout.java (hunk 4 above)
Comment (rschlussel) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PickTableLayout.java]: see comments above.  We want to be able to run this rule more than once before AddExchanges, but the invocation after AddExchanges needs to not overwrite the layout chosen by AddExchanges.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PickTableLayout.java [hunk 9]
```diff
@@ -78,30 +82,26 @@ private PickTableLayoutForPredicate(Metadata metadata)
...
-            if (rewrittenTableScan instanceof TableScanNode || rewrittenTableScan instanceof ValuesNode || (((FilterNode) rewrittenTableScan).getPredicate() != filterNode.getPredicate())) {
-                return Optional.of(rewrittenTableScan);
-            }
-
-            return Optional.empty();
+            TableLayoutRewriter tableLayoutRewriter = new TableLayoutRewriter(metadata, context.getSession(), context.getSymbolAllocator(), context.getIdAllocator(), sqlParser);
+            return tableLayoutRewriter.planTableScan((TableScanNode) source, filterNode.getPredicate());
         }
 
         private boolean shouldRewriteTableLayout(TableScanNode source)
```
[the remark this answers is not in the document]
Comment (rschlussel) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PickTableLayout.java]: you do need to check `!source.getLayout().isPresent()` because when replaceExistingLayout is false, you still want to add a layout if none existed.  You can't just check `planChanged` then because you only want to return the changed plan if the original plan didn't have a layout.

[the remark this answers is not in the document]
Comment (sopel39) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PickTableLayout.java]: True, but `source.getOriginalConstraint() == BooleanLiteral.TRUE_LITERAL;` can be removed

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TableLayoutRewriter.java (hunk 8 above)
Comment (rschlussel) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TableLayoutRewriter.java]: see https://github.com/Teradata/presto/pull/592 and in particular https://github.com/Teradata/presto/pull/592#discussion_r124269935

Comment (sopel39) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TableLayoutRewriter.java]: Sure, it might change after one execution of `PickTableLayout` rule, especially since the predicate is rewritten by:
```
        Expression resultingPredicate = combineConjuncts(
                decomposedPredicate.getRemainingExpression(),
                DomainTranslator.toPredicate(layout.getUnenforcedConstraint().transform(assignments::get)));
```
, but the second call to the rule (when nothing changed) should produce exactly the same predicate (it should be deterministic). I am good either way.

File: presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestPickTableLayout.java [hunk 10]
```diff
@@ -46,13 +49,15 @@
...
+    private RuleSet pickTableLayoutReplaceExisting;
     private TableHandle nationTableHandle;
     private TableLayoutHandle nationTableLayoutHandle;
 
     @BeforeMethod
     public void setUpPerMethod()
     {
-        pickTableLayout = new PickTableLayout(tester().getMetadata());
+        pickTableLayout = new PickTableLayout(tester().getMetadata(), new SqlParser(), false);
+        pickTableLayoutReplaceExisting = new PickTableLayout(tester().getMetadata(), new SqlParser(), true);
```
Comment (sopel39) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestPickTableLayout.java]: TPCH connector does support layouts for `orderstatus` column in `orders` table. Therefore you can add proper tests with connector accepting predicate fully, see: https://github.com/prestodb/presto/pull/8865/commits/da183f4b4300c708a4f321173c88d72594687b90 and https://github.com/prestodb/presto/pull/8865/commits/e0f2750bcbe6aedb16cca559ab3727cc8853826f

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TableLayoutRewriter.java (hunk 8 above)
Comment (rschlussel) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TableLayoutRewriter.java]: Leaving this as is.  It was introduced to fix an infinite loop where predicates were reordered with every iteration.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TableLayoutRewriter.java [hunk 11]
```diff
@@ -80,12 +84,12 @@ public PlanNode planTableScan(TableScanNode node, Expression predicate)
                 .filter(layoutHasAllNeededOutputs(node))
                 .collect(toImmutableList());
         if (layouts.isEmpty()) {
-            return new ValuesNode(idAllocator.getNextId(), node.getOutputSymbols(), ImmutableList.of());
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TableLayoutRewriter.java]: Shouldn't you make `node.getCurrentConstrait()` part of a layout query? What about following scenario:
```
FilterNode(a IN (1, 2))
 TableScanNode(currentConstrait=none)
```
then first `PickTableLayout` rule would produce
```
FilterNode(a IN (2))
 TableScanNode(currentConstrait=(a IN (1))
```
but the 3rd call of `PickTableLayout` rule would produce:
```
ValuesNode(rows=[])
```
since intersect here: https://github.com/prestodb/presto/pull/8735/commits/0cc4b665b05afed5e08f24b44d037268d7c16dbc#diff-80422a11a5eab4c4e0e29dc080660e84R77 would produce empty `TupleDomain`.

Comment (rschlussel) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TableLayoutRewriter.java]: It would be incorrect to push down `a in (1)` to the tablescan.  `a in(1,2)` is basically an or `a = 1 OR a =2`.  So you couldn't filter by just one of them.

**Comment by rschlussel** [comment]
@kokosing @sopel39 -- addressed comments. 
@sopel39 I added your commits too for adding a fully enforced test.

File: presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestPickTableLayout.java [hunk 12]
```diff
@@ -60,7 +63,7 @@
     @BeforeClass
     public final void setUp()
     {
-        tester = new RuleTester();
+        tester = new RuleTester(true);
```
Comment (kokosing) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestPickTableLayout.java]: maybe instead of this you could make `RuleTester` to accept `QueryRunner` in constructor, then `BaseRuleTest` could have a protected method to create query runner, which could be overwritten by this test. 
What do you think?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java (hunk 1 above)
Comment (kokosing) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java]: > cost-based optimizers can go anywhere

Are you sure we want to go with such approach? There might be more things than picking table layout which may influence stats quality (like `SimplifyExpressions`). I would prefer to prepare for now a single place, where we know that stats should be good, and there put all the cost-based related rules. How does it sound to you?

Moreover we may want to to put a `checkState` to stats estimation code to prevent running them without a layout. Just in case someone wanted to put cost-related rule into a wrong place.

Comment (rschlussel) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java]: What about cost-based optimizations that could enable choosing a different layout.  I guess we could worry about that when we have such an optimization.

I like the idea of a `checkState` in the stats code.

File: presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestPickTableLayout.java (hunk 12 above)
Comment (rschlussel) [reply presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestPickTableLayout.java]: tests that need to do other kinds of set up become more complicated because they don't extend BaseRuleTest, so then you need to create the sessions/QueryRunners for those tests separately there.  This applies to `TestTransformExistsApplyToScalaraLateralJoin` and `TestTransforCorrelatedScalarAggregationToJoin`.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PickTableLayout.java (hunk 4 above)
Comment (rschlussel) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PickTableLayout.java]: realized that we don't need to run after add exchanges, so I removed the boolean parameter

Comment (kokosing) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PickTableLayout.java]: instead of removing this flag in a subsequent commit, please do not introduce it all.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PickLayout.java [hunk 13]
```diff
@@ -1,90 +0,0 @@
...
-
-import java.util.Map;
-
-import static java.util.Objects.requireNonNull;
-
-/**
- * Pick an arbitrary layout if none has been chosen
- */
-@Deprecated
-public class PickLayout
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PickLayout.java]: Please leave this class for a while

Comment (rschlussel) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PickLayout.java]: but it's not used anymore.
It can't be the non-iterative alternative to PickTableLayout because it only gets run once, so you can't run it too early or even later if there's a table layout that's chosen by something else earlier (e.g. PickTableLayout).

**Comment by rschlussel** [comment]
commits squashed

File: presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestPickTableLayout.java [hunk 14]
```diff
@@ -40,40 +42,58 @@
...
+    {
+        closeAllRuntimeException(tester);
+        tester = null;
+    }
 
     @BeforeMethod
     public void setUpPerMethod()
     {
-        pickTableLayout = new PickTableLayout(tester().getMetadata(), new SqlParser());
+        pickTableLayout = new PickTableLayout(tester.getMetadata(), new SqlParser());
```
Comment (findepi) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestPickTableLayout.java]: Instead of defining all following tests' code, please define 
```
private RuleTester tester()
    {
        return tester;
    }
```

File: presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestPickTableLayout.java [hunk 15]
```diff
@@ -40,40 +42,58 @@
...
 
-        ConnectorId connectorId = tester().getCurrentConnectorId();
+        ConnectorId connectorId = tester.getCurrentConnectorId();
         nationTableHandle = new TableHandle(
                 connectorId,
                 new TpchTableHandle(connectorId.toString(), "nation", 1.0));
+        ordersTableHandle = new TableHandle(
+                connectorId,
+                new TpchTableHandle(connectorId.toString(), "orders", 1.0));
```
Comment (findepi) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestPickTableLayout.java]: This change is unrelated to "Use RuleTester instance explicitly in PickTableLayout tests" commit.

File: presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestPickTableLayout.java [hunk 16]
```diff
@@ -40,40 +42,58 @@
...
 public class TestPickTableLayout
-        extends BaseRuleTest
 {
     private static final TableLayoutHandle DUMMY_TABLE_LAYOUT_HANDLE = new TableLayoutHandle(new ConnectorId("tpch"),
             TestingTransactionHandle.create(),
             new TpchTableLayoutHandle(new TpchTableHandle("local", "nation", 1.0), Optional.empty()));
 
     private RuleSet pickTableLayout;
     private TableHandle nationTableHandle;
+    private TableHandle ordersTableHandle;
```
Comment (findepi) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestPickTableLayout.java]: This change is unrelated to "Use RuleTester instance explicitly in PickTableLayout tests" commit.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PickLayout.java (hunk 13 above)
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PickLayout.java]: This commit belongs before "Add correctness logic from AddExchanges to TableLayoutRewriter" (the last reference to this class was lost in that commit, but it should not be mixed with other changes introduced there). 

@kokosing are you still concerned we're not ready to remove this class just yet?

**Comment by findepi** [comment]
@martint are you OK with removing the old `PickLayout` optimizer?

**Comment by findepi** [comment]
@sopel39 i vaguely remember there was a topic of `PickTableLayout` being able to fire infinitely. Do you remember the details?

**Comment by sopel39** [comment]
Yes. Check Martin PR removing legacy optimizer rules. There is mine comment there with link to commit that catches infinite rule execution.

**Comment by findepi** [comment]
Do you remember where was it?

**Comment by sopel39** [comment]
https://github.com/prestodb/presto/pull/9424#issuecomment-347460210. It could be that Rebecca fixed the issue idk

**Comment by kokosing** [comment]
> are you OK with removing the old PickLayout optimizer?

If possible I would prefer to keep the old optimizer for a while

**Comment by findepi** [comment]
> https://github.com/prestodb/presto/pull/9424#issuecomment-347460210. It could be that Rebecca fixed the issue idk

@sopel39 would it be possible to have a test reproducing the problem?

**Comment by sopel39** [comment]
Check this branch https://github.com/Teradata/presto/commits/ks/rule_apply_iterative

**Comment by kokosing** [comment]
Superseded by #9701
